cachix/src/Cachix/Client/Push.hs

```diff
@@ -140,133 +161,66 @@ pushSingleStorePath ::
+pushSingleStorePath pushParams storePath = retryAll $ \retrystatus -> do
   storeHash <- liftIO $ Store.getStorePathHash storePath
-  let name = pushParamsName cache
-      strategy = pushParamsStrategy cache storePath
-  -- Check if narinfo already exists
-  res <-
-    liftIO $
-      retryHttp $
-        (`runClientM` pushParamsClientEnv cache) $
-          API.narinfoHead
-            cachixClient
-            (getCacheAuthToken (pushParamsSecret cache))
-            name
-            (NarInfoHash.NarInfoHash (decodeUtf8With lenientDecode storeHash))
+  let strategy = pushParamsStrategy pushParams storePath
+  res <- liftIO $ narinfoExists pushParams storeHash
   case res of
     Right NoContent -> onAlreadyPresent strategy -- we're done as store path is already in the cache
     Left err
-      | isErr err status404 -> uploadStorePath cache storePath retrystatus
+      | isErr err status404 -> uploadStorePath pushParams storePath retrystatus
       | isErr err status401 -> on401 strategy err
       | otherwise -> onError strategy err
 
+narinfoExists :: PushParams m r -> ByteString -> IO (Either ClientError NoContent)
+narinfoExists pushParams storeHash = do
+  let cacheName = pushParamsName pushParams
+      authToken = getCacheAuthToken (pushParamsSecret pushParams)
+  retryHttp $
+    (`runClientM` pushParamsClientEnv pushParams) $
+      API.narinfoHead
+        cachixClient
+        authToken
+        cacheName
+        (NarInfoHash.NarInfoHash (decodeUtf8With lenientDecode storeHash))
+
 getCacheAuthToken :: PushSecret -> Token
 getCacheAuthToken (PushToken token) = token
 getCacheAuthToken (PushSigningKey token _) = token
 
 uploadStorePath ::
-  (MonadUnliftIO m) =>
+  (MonadUnliftIO m, MonadCatch m) =>
   -- | details for pushing to cache
   PushParams m r ->
   StorePath ->
   RetryStatus ->
   -- | r is determined by the 'PushStrategy'
   m r
-uploadStorePath cache storePath retrystatus = do
-  let store = pushParamsStore cache
+uploadStorePath pushParams storePath retrystatus = do
+  let store = pushParamsStore pushParams
+      strategy = pushParamsStrategy pushParams storePath
+
   -- TODO: storePathText is redundant. Use storePath directly.
   storePathText <- liftIO $ Store.storePathToPath store storePath
-  let (storeHash, storeSuffix) = splitStorePath $ toS storePathText
-      cacheName = pushParamsName cache
-      authToken = getCacheAuthToken (pushParamsSecret cache)
-      clientEnv = pushParamsClientEnv cache
-      strategy = pushParamsStrategy cache storePath
-      withCompressor = case compressionMethod strategy of
-        BinaryCache.XZ -> defaultWithXzipCompressorWithLevel (compressionLevel strategy)
-        BinaryCache.ZSTD -> defaultWithZstdCompressorWithLevel (compressionLevel strategy)
-      cacheClientEnv =
-        clientEnv
-          { baseUrl = (baseUrl clientEnv) {baseUrlHost = toS cacheName <> "." <> baseUrlHost (baseUrl clientEnv)}
```

Comment:
Is there any good reason for routing API calls via the cache subdomain? Doesn't seem necessary to me. Am I missing something?